app/scripts/lib/WeakRefObjectMap.ts

```diff
@@ -0,0 +1,74 @@
+type WeakRefCompatibleObject<T> = {
+  [P in keyof T]: T[P] extends object ? WeakRef<T[P]> : never;
+};
+
+export class WeakRefObjectMap {
+  map: Map<string, WeakRefCompatibleObject<Record<string, object>>>;
+
+  constructor() {
+    this.map = new Map();
+  }
+
+  set(key: string, value: Record<string, object>) {
+    const weakReffedValueObj = {} as WeakRefCompatibleObject<
+      Record<string, object>
+    >;
+    Object.keys(value).forEach((valueKey) => {
+      weakReffedValueObj[valueKey] = new WeakRef(value[valueKey]);
+    });
+    this.map.set(key, weakReffedValueObj);
+  }
+
+  get(key: string) {
+    const weakReffedValue = this.map.get(key);
+    if (!weakReffedValue) {
+      return undefined;
+    }
+
+    const deReffedValue = {} as { [dereffedValueObjKey: string]: WeakKey };
+    for (const weakReffedValueKey of Object.keys(weakReffedValue)) {
+      const deref = weakReffedValue[weakReffedValueKey].deref();
+      if (deref === undefined) {
+        this.map.delete(weakReffedValueKey);
+        return undefined;
+      }
+      deReffedValue[weakReffedValueKey] = deref;
+    }
+
+    return deReffedValue;
+  }
+
+  has(key: string) {
+    return this.get(key) !== undefined;
+  }
+
+  delete(key: string) {
+    const value = this.get(key);
+    if (value !== undefined) {
+      return this.map.delete(key);
+    }
+    return false;
+  }
+
+  clear() {
+    this.map.clear();
+  }
+
+  forEach(
+    callback: (
+      value: WeakRefCompatibleObject<Record<string, object>>,
+      key: string,
+      map: Map<string, WeakRefCompatibleObject<Record<string, object>>>,
+    ) => void,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    thisArg?: any,
+  ): void {
+    this.map.forEach((value, key) => {
+      if (thisArg) {
+        callback.call(thisArg, value, key, this.map);
+      } else {
+        callback(value, key, this.map);
+      }
+    });
```

Comment:
should this be calling `this.get()` and skipping undefineds?

Reply:
> should this be calling this.get()
Probably...? Though you'll notice that in practice we're only referencing the key/domain where we call `forEach`: https://github.com/MetaMask/core/pull/4130/files#diff-5e705302d595033f24cac5f20d14ecf307709fd61c46a6d01bf9519bf5b91d86R217

Reply:
But I can make this update

Reply:
Actually I don't really think it makes sense to get the dereffed instance, especially since it could end up mutating the collection if it finds that a weak ref has been garbage collected. I think it could lead to unpredictable behavior, and its easy enough to go in and deref in callback if you need to.

Reply:
To clarify, currently this loop over `this.map` will call `callback` for all values in the map, including values that have been garbage collected. Additionally, `callback` will be provided the WeakRef'ed object. My assumption would be that all WeakRef wrapping and unwrapping would be hidden from the consumer of `WeakRefObjectMap`. To the consumer, this class would work like a regular Map, except that entries could disappear

Reply:
I do see that we only use the key/domain, but even in that case, should we be iterating over entries where the value has been garbage collected already?